README
NAME
    "Text::Categorize::Textrank" - Computes textrank of text.

SYNOPSIS
      use Text::Categorize::Textrank;
      use Data::Dump qw(dump);
      my $textranker = Text::Categorize::Textrank->new;

DESCRIPTION
    "Text::Categorize::Textrank" uses the modules Text::StemTagPOS and
    Graph::Centrality::Pagerank to compute the textrank of words given the
    text. Encoding of all text should be in Perl's internal format; see the
    module Text::Iconv for converting text from various encodings.

CONSTRUCTOR
  "new"
    The method "new" creates an instance of the "Text::Categorize::Textrank"
    class with the following parameters:

    "isoLangCode"
         isoLangCode => 'en'

        "isoLangCode" is the ISO language code of the language that will be
        tagged and stemmed by the object. It must be 'en', which is also the
        default. In the future 'de' may be added.

    "endingSentenceTag"
         endingSentenceTag => 'PP'

        "endingSentenceTag" is the part-of-speech tag that should be used it
        indicate the end of a sentence. The default is 'PP'. The value of
        this tag must be a tag generated by the module Lingua::EN::Tagger.

    "listOfPOSTypesToKeep"
         listOfPOSTypesToKeep => [...]

        The Textrank method preprocesses the text so that only specific
        parts of speech are retained and used to build the graph
        representing the text. The module Lingua::EN::Tagger is used to tag
        the parts-of-speech of the text. The parts-of-speech retained can be
        specified by word types, where the type is a combination of 'ALL',
        'ADJECTIVES', 'ADVERBS', 'CONTENT_WORDS', 'NOUNS', 'PUNCTUATION',
        'TEXTRANK_WORDS', or 'VERBS'. The default is "[qw(TEXTRANK_WORDS)]",
        which equates to "[qw(ADJECTIVES NOUNS)]".

    "listOfPOSTagsToKeep"
         listOfPOSTagsToKeep => [...]

        "listOfPOSTagsToKeep" provides finer control over the
        parts-of-speech to be retained when filtering previously tagged
        text. For a list of all the possible tags call
        "getListOfPartOfSpeechTags".

    "stdevFactor"
         stdevFactor => 1

        If "avgPageRank" is the average pagerank score and "stdevPageRank"
        is their standard deviation, then a words page rank must be greater
        than or equal to the "avgPageRank + stdevFactor * stdevPageRank" to
        be selected as keyword. The default is one.

    "computeStatsIgnoringRepeatedWords"
         computeStatsIgnoringRepeatedWords => 0

        When computing the average and standard deviation of the pageranks
        of the words this can be done by treating the text as a sequence of
        words or a set of words; in the latter case repeating words are
        ignored. For example, suppose the only words in the text were "very"
        and "good" with pageranks of .3 and .7 respectively. If the original
        text is "Very very good.", then the average pagerank of the text is
        "(.3 + .3 + .7) / 3". If the pagerank is computed ignoring repeated
        words it is "(.3 + .7) / 2".

    "getTopKeyPhrases"
         getTopKeyPhrases => 10

        Instead of using "stdevFactor" to chose the top key phrases to
        extract from the text, "getTopKeyPhrases" can be used to return the
        top scoring words or phrases. This is the default method, with the
        top 10 words or phrases being returned. Note, if the text has ten
        words or less, the entire text will be returned as one phrase.

METHODS
  "getTextrankScores"
     getTextrankScores (...)

    The method "getTextrankScores" returns a hash-reference of all the
    stemmed words used in computing the textrank. The value of each hash is
    the textrank of the word. The sum of all the textranks is one.

    "listOfText"
         listOfText => [...]

        "listOfText" is the array reference of the strings of text that is
        to be categorized.

    "tokenEdgeSpan"
         tokenEdgeSpan => 3

        When building the textrank graph, "tokenEdgeSpan" tokenEdgeSpan
        holds the number of successive words linked by an edge. For example,
        given the tokens "apple orange pear", the edges "[apple, orange]"
        and "[apple, pear]" will be created for the token "apple" if
        $tokenEdgeSpan is three. The default is two. Need to change this!

    "undirectedGraph"
         directedGraph => 0

        Computing the textrank of the works involves building a graph from
        the words. If "directedGraph" evaluates to true, then the graph will
        be directed. The default is false, that is, the graph is undirected.

    "pageRankDampeningFactor"
         pageRankDampeningFactor => 0.85

        When computing the pagerank of the textrank graph, the dampening
        factor specified by "pageRankDampeningFactor" will be used; it
        should range from zero to one. The default is 0.85.

    "useEdgeWeights"
         useEdgeWeights => 1

        If "useEdgeWeights" is true, then when computing the pagerank of the
        textrank graph the edge weights will be included in the calculation.
        The default is true.

    "addEdgesSpanningSentences"
         addEdgesSpanningSentences => 1

        If "addEdgesSpanningSentences" is true, then when build the textrank
        graph, links between the words at the end of sentence and the
        beginning of the next sentence will be made. The default is true.

INSTALLATION
    To install the module run the following commands:

      perl Makefile.PL
      make
      make test
      make install

    If you are on a windows box you should use 'nmake' rather than 'make'.

AUTHOR
     Jeff Kubina<jeff.kubina@gmail.com>

COPYRIGHT
    Copyright (c) 2009 Jeff Kubina. All rights reserved. This program is
    free software; you can redistribute it and/or modify it under the same
    terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.

SEE ALSO
    Lingua::Stem::Snowball, Lingua::EN::Tagger, and Lingua::DE::Tagger



Changes
Revision history for Perl module Text::Categorize::Textrank

0.50  8 Jan 2012
  - released to CPAN.


lib/Text/Categorize/Textrank.pm
package Text::Categorize::Textrank;

use strict;
use warnings;
use Graph;
use Graph::Centrality::Pagerank;
use Data::Dump qw(dump);

BEGIN {
    use Exporter ();
    use vars qw($VERSION @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);
    $VERSION     = '0.50';
    @ISA         = qw(Exporter);
    @EXPORT      = qw(getTextrankOfListOfTokens);
    @EXPORT_OK   = qw(getTextrankOfListOfTokens);
    %EXPORT_TAGS = ();
}

#12345678901234567890123456789012345678901234
#Method to rank potential keywords of text.

=head1 NAME

C<Text::Categorize::Textrank> - Method to rank potential keywords of text.

=head1 SYNOPSIS

	use strict;
	use warnings;
	use Text::Categorize::Textrank;
	use Data::Dump qw(dump);
	my $listOfTokens = [ [qw(This is the first sentence)], [qw(Here is the second sentence)] ];
	my $hashOfTextrankValues = getTextrankOfListOfTokens(listOfTokens => $listOfTokens);
	dump $hashOfTextrankValues;

=head1 DESCRIPTION

C<Text::Categorize::Textrank> provides a routine for ranking the words in
text as potential keywords. It implements a version of the textrank algorithm
from the report I<TextRank: Bringing Order into Texts> by R. Mihalcea and P. Tarau.

=head1 ROUTINES

=head2 C<getTextrankOfListOfTokens>

The routine C<getTextrankOfListOfTokens> returns a hash reference containing the textrank value
for all the tokens in the lists provided; the textrank values sum to one. The textrank of a token
is its pagerank in the graph obtained by joining neighboring tokens with an edge, called the
token graph.

Usually, C<listOfTokens> should I<not> be applied to all the words of the text. The
complete textrank algorithm first filters the words in the text to only the nouns and adjectives.
See L<Text::Categorize::Textrank::En> to compute the textrank of English text.

=over

=item C<listOfTokens>

  listOfTokens => [[...], [...], ...[...]]

C<listOfTokens> is an array reference containing the list of tokens that are
to be ranked using textrank. Each list is also an array reference of tokens that should
correspond to the list of tokens in a sentence. For example,
C<[[qw(This is the first sentence)], [qw(Here is the second sentence)]]>.

=item C<edgeCreationSpan>

  edgeCreationSpan => 1

For each token in the C<listOfTokens>, C<edgeCreationSpan> is the number of successive
tokens used to make an edge in the token graph. For example, if
C<edgeCreationSpan> is two, then given the token sequence C<"apple orange pear">
the edges C<[apple, orange]> and C<[apple, pear]> will be added to the token
graph for the token C<apple>. The default is one.

Note that loop edges are ignored. For example,
if C<edgeCreationSpan> is two, then given the token sequence C<"daba daba doo">
the edge C<[daba, daba]> is disguarded but the edge C<[daba, doo]> is
added to the token graph.

=item C<directedGraph>

  directedGraph => 0

If C<directedGraph> is true, the textranks
are computed from the directed token graph, if false, they
are computed from the undirected version of the graph. The default is false.

=item C<dampeningFactor>

  dampeningFactor => 0.85

When computing the textranks of the token graph, the dampening factor
specified by C<dampeningFactor> will
be used; it should range from zero to one. The default is 0.85.

=begin html

The Wikipedia article on <a href="http://en.wikipedia.org/wiki/PageRank">pagerank</a> has a good explaination of the
<a href="http://en.wikipedia.org/wiki/PageRank#Damping_factor">dampening factor</a>.<br>&nbsp;

=end html

=item C<addEdgesSpanningLists>

  addEdgesSpanningLists => 1

If C<addEdgesSpanningLists> is true, then when building the token graph, links
between the tokens at the end of a list and the beginning of the next list
will be made. For example, for the lists C<[[qw(This is the first list)], [qw(Here is the second list)]]>
the edge C<[list, Here]> will be added to the token graph. The default is true.

=item C<tokenWeights>

  tokenWeights => {}

C<tokenWeights> is an optional hash reference that can provide a weight for a subset
of the tokens provided by C<listOfTokens>. If C<tokenWeights> is not defined for any token
in C<listOfTokens>, then each
token has a weight of one. If C<tokenWeights> is defined for
at least one node in the graph, then the default weight of any undefined
token is zero.

=back

=cut

sub getTextrankOfListOfTokens
{
  my %Parameters = @_;

  # if there are no lists, return undef.
  return undef unless exists $Parameters{listOfTokens};
  my $listOfTokens = $Parameters{listOfTokens};

  # edgeCreationSpan holds the number of sucessive tokens linked by an edge for each token.
  # for example, given tokens "apple orange pear", the edges [apple, orange] and [apple, pear]
  # will be created for token apple if $edgeCreationSpan is two.
  my $edgeCreationSpan = 1;
  $edgeCreationSpan = abs $Parameters{edgeCreationSpan} if (exists ($Parameters{edgeCreationSpan}));
  $edgeCreationSpan = 1 if ($edgeCreationSpan < 1);

  # set the type of graph built from the tokens to compute the textrank.
  $Parameters{undirected} = 1;
  $Parameters{undirected} = !$Parameters{directedGraph} if exists ($Parameters{directedGraph});

  # get the pagerank dampening factor.
  $Parameters{dampeningFactor} = 0.85 unless exists ($Parameters{dampeningFactor});
  $Parameters{dampeningFactor} = abs $Parameters{dampeningFactor};
  $Parameters{dampeningFactor} = 1 if ($Parameters{dampeningFactor} > 1);

  # set the addEdgesSpanningLists flag.
  my $addEdgesSpanningLists = 1;
  $addEdgesSpanningLists = $Parameters{addEdgesSpanningLists} if exists $Parameters{addEdgesSpanningLists};

  # build the graph of the links between tokens.
  my $graph = Graph->new(directed => !$Parameters{undirected});

  # if edges will span lists, make listOfTokens into just one list.
  if ($addEdgesSpanningLists)
  {
    my @allWords;
    foreach my $list (@$listOfTokens)
    {
      push @allWords, @$list;
    }
    $listOfTokens = [\@allWords];
  }

  # add all the edges to the graph.
  foreach my $list (@$listOfTokens)
  {
    # get the total tokens in the list.
    my $totalTokens = $#$list + 1;
    for (my $i = 0; $i < $totalTokens; $i++)
    {
      # since isolalted vertices are possible, add the vertex.
      # for example, if a list had only one token in it, it would
      # not be linked to any other token.
      $graph->add_vertex ($list->[0]);

      # get the index of the last token to link to the current token in
      # the list.
      my $lastToken = $i + $edgeCreationSpan + 1;
      $lastToken = $totalTokens if ($lastToken > $totalTokens);
      for (my $j = $i + 1; $j < $lastToken; $j++)
      {
        # skip loop edges.
        next if ($list->[$i] eq $list->[$j]);

        # create the edge.
        my @edge = ($list->[$i], $list->[$j]);

        # if the edge exists already, add to its weight.
        if ($graph->has_edge (@edge))
        {
          my $weight = $graph->get_edge_weight (@edge);
          $weight = 1 unless defined $weight;
          ++$weight;
          $graph->add_weighted_edge (@edge, $weight);
        }
        else
        {
          # edge does not exist, default edge weight is one.
          $graph->add_weighted_edge (@edge, 1);
        }
      }
    }
  }

  # get the pagerank computer.
  my $pageRankEngine = Graph::Centrality::Pagerank->new ();

  # set the parameter for inclusion of edge weights.
  $Parameters{useEdgeWeights} = 1;

  # set the node weights if provided.
  $Parameters{nodeWeights} = $Parameters{tokenWeights} if ((exists $Parameters{tokenWeights}) && (defined $Parameters{tokenWeights}));

  # compute and return the pagerank of the tokens.
  return $pageRankEngine->getPagerankOfNodes (%Parameters, graph => $graph);
}

=head1 INSTALLATION

To install the module run the following commands:

  perl Makefile.PL
  make
  make test
  make install

If you are on a windows box you should use 'nmake' rather than 'make'.

=head1 BUGS

Please email bugs reports or feature requests to C<bug-text-categorize-textrank@rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Text-Categorize-Textrank>.  The author
will be notified and you can be automatically notified of progress on the bug fix or feature request.

=head1 AUTHOR

 Jeff Kubina<jeff.kubina@gmail.com>

=head1 COPYRIGHT

Copyright (c) 2009 Jeff Kubina. All rights reserved.
This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=head1 KEYWORDS

categorize, keywords, keyphrases, nlp, pagerank, textrank

=head1 SEE ALSO

=begin html

This package implements the Textrank algorithm from the report
<a href="http://bit.ly/akSJok">TextRank: Bringing Order into Texts</a>
by <a href="http://www.cse.unt.edu/~rada/">Rada Mihalcea</a> and <a href="www.cse.unt.edu/~tarau/">Paul Tarau</a>;
which is related to <a href="http://en.wikipedia.org/wiki/PageRank">pagerank</a>.

=end html

L<Graph>, L<Graph::Centrality::Pagerank>, L<Log::Log4perl>, L<Text::Categorize::Textrank::En>

=cut

1;
# The preceding line will help the module return a true value


lib/Text/Categorize/Textrank/En.pm
package Text::Categorize::Textrank::En;

use strict;
use warnings;
use Log::Log4perl;
use Text::StemTagPOS;
use Text::Categorize::Textrank;
use Data::Dump qw(dump);

# TODO: need parameter for maximum phrase length.

BEGIN {
    use Exporter ();
    use vars qw($VERSION @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);
    $VERSION     = '0.50';
    @ISA         = qw(Exporter);
    @EXPORT      = qw(getTextrankInfoOfText);
    @EXPORT_OK   = qw(getTextrankInfoOfText);
    %EXPORT_TAGS = ();
}

#12345678901234567890123456789012345678901234
#Find potential keywords in English text.

=head1 NAME

C<Text::Categorize::Textrank::En> - Find potential keywords in English text.

=head1 SYNOPSIS

	use strict;
	use warnings;
	use Text::Categorize::Textrank::En;
	use Data::Dump qw(dump);
	my $textrankerEn = Text::Categorize::Textrank::En->new();
	my $text         = 'This is the first sentence. Here is the second sentence.';
	my $results      = $textrankerEn->getTextrankInfoOfText(listOfText => [$text]);
	dump $results->{hashOfTextrankValues};

=head1 DESCRIPTION

C<Text::Categorize::Textrank::En> provides methods for ranking the words in English
text as potential keywords. It implements a version of the textrank algorithm
from the report I<TextRank: Bringing Order into Texts> by R. Mihalcea and P. Tarau.

Encoding of all text should be in Perl's internal format; see L<Text::Iconv> or L<Encode> for
converting text from various encodings.

=head1 CONSTRUCTOR

=head2 C<new>

The method C<new> creates an instance of the C<Text::Categorize::Textrank::En>
class with the following parameters:

=over

=item C<endingSentenceTag>

 endingSentenceTag => 'PP'

C<endingSentenceTag> is the part-of-speech tag that should be used to indicate
the end of a sentence. The default is 'PP'. The value of this tag must be
a tag generated by the module L<Lingua::EN::Tagger>.

=item C<listOfPOSTypesToKeep>

 listOfPOSTypesToKeep => [qw(TEXTRANK_WORDS)]

The textrank algorithm preprocesses the text so that only certain parts-of-speech (POS) are retained
and used to build the graph representing the text. The module L<Lingua::EN::Tagger> is used
to tag the parts-of-speech of the text. The parts-of-speech retained can be specified by
word types, where the type is a combination of 'ALL', 'ADJECTIVES', 'ADVERBS', 'CONTENT_WORDS', 'NOUNS', 'PUNCTUATION',
'TEXTRANK_WORDS', or 'VERBS'. The default is C<[qw(TEXTRANK_WORDS)]>, which equates to
C<[qw(ADJECTIVES NOUNS)]>.

=item C<listOfPOSTagsToKeep>

 listOfPOSTagsToKeep => [...]

C<listOfPOSTagsToKeep> provides finer control over the
parts-of-speech to be retained when filtering the tagged text. For a list
of all the possible tags call C<getListOfPartOfSpeechTags()>.

=back

=cut

sub new
{
  my ($Class, %Parameters) = @_;
  my $Self = bless ({}, ref ($Class) || $Class);

  # get the POS/stemmer engine.
  $Self->{posTaggerStemmerEngine} = Text::StemTagPOS->new (%Parameters);

  return $Self;
}

=head1 METHODS

=head2 C<getTextrankInfoOfText>

 getTextrankInfoOfText (...)

The method C<getTextrankInfoOfText> returns a data structure (hash-reference)
containing all the stemmed
words partitioned into their sentences (L<listOfStemmedTaggedSentences|Text::StemTagPOS/getStemmedAndTaggedText>), the subset of words
used to compute the textranks (L<listOfFilteredSentences|Text::StemTagPOS/getTaggedTextToKeep>), and the textrank
of the tokens (L<hashOfTextrankValues|Text::Categorize::Textrank/getTextrankOfListOfTokens>) that occur in C<listOfFilteredSentences>. The sum of all the textrank
values is one.

More precisely, if C<$results> is the returned hash, then C<$results-E<gt>{listOfStemmedTaggedSentences}>
contains the array reference generated by the L<getStemmedAndTaggedText|Text::StemTagPOS/getStemmedAndTaggedText>
method of L<Text::StemTagPOS>, C<$results-E<gt>{listOfFilteredSentences}>
contains the array reference generated by L<getTaggedTextToKeep|Text::StemTagPOS/getTaggedTextToKeep>
of L<Text::StemTagPOS>, and C<$results-E<gt>{hashOfTextrankValues}>
holds the hash of the textrank values computed by L<getTextrankOfListOfTokens|Text::Categorize::Textrank/getTextrankOfListOfTokens>.
C<$results-E<gt>{useStemmedWords}> is also set to the value of C<useStemmedWords>.

=over

=item C<listOfStemmedTaggedSentences>

 listOfStemmedTaggedSentences => [...]

C<listOfStemmedTaggedSentences> is the array reference containing the list of stemmed and part-of-speech
tagged sentences from L<Text::StemTagPos>. If C<listOfStemmedTaggedSentences> is not defined, then the
text to be processed should be provided via C<listOfText>.

=item C<listOfText>

 listOfText => [...]

C<listOfText> is an array reference containing the strings of text to be categorized. C<listOfText> is
only used if C<listOfStemmedTaggedSentences> is undefined.

=item C<edgeCreationSpan>

  edgeCreationSpan => 1

For each word in the text, C<edgeCreationSpan> is the number of successive
words used to make an edge in the textrank token graph. For example, if
C<tokenEdgeSpanSize> is two, then given the word sequence C<"apple orange pear">
the edges C<[apple, orange]> and C<[apple, pear]> will be added to the text
graph for the word C<apple>. The default is one.

Note that loop edges are ignored. For example,
if C<edgeCreationSpan> is two, then given the word sequence C<"daba daba doo">
the edge C<[daba, daba]> is disguarded but the edge C<[daba, doo]> is
added to the token graph.

=item C<directedGraph>

  directedGraph => 0

If C<directedGraph> is true, the textranks
are computed from the directed token graph, if false, they
are computed from the undirected version of the graph. The default is false.

=item C<pageRankDampeningFactor>

  pageRankDampeningFactor => 0.85

When computing the textranks of the token graph, the dampening factor
specified by C<pageRankDampeningFactor> will
be used; it should range from zero to one. The default is 0.85.

=begin html

The Wikipedia article on <a href="http://en.wikipedia.org/wiki/PageRank">pagerank</a> has a good explaination of the
<a href="http://en.wikipedia.org/wiki/PageRank#Damping_factor">dampening factor</a>.<br>&nbsp;

=end html

=item C<addEdgesSpanningSentences>

  addEdgesSpanningLists => 1

If C<addEdgesSpanningLists> is true, then when building the token graph, links
between the tokens at the end of a list and the beginning of the next list
will be made. For example, for the lists C<[[qw(This is the first list)], [qw(Here is the second list)]]>
the edge C<[list, Here]> will be added to the token graph. The default is true.

=item C<useStemmedWords>

  useStemmedWords => 1

If C<useStemmedWords> is true, then when building the token graph, the stemmed
words are used as the id of each node, otherwise the original words
are used; in both cases the stemmed or original words are converted to
lowercase. The default is true.

=back

=cut

sub getTextrankInfoOfText
{
  my ($Self, %Parameters) = @_;

  # get the text to process.
  my $listOfStemmedTaggedSentences;
  if (exists ($Parameters{listOfStemmedTaggedSentences}))
  {
    $listOfStemmedTaggedSentences = $Parameters{listOfStemmedTaggedSentences};
  }
  elsif (exists($Parameters{listOfText}))
  {
    $listOfStemmedTaggedSentences = $Self->{posTaggerStemmerEngine}->getStemmedAndTaggedText ($Parameters{listOfText});
  }
  else
  {
    my $logger = Log::Log4perl->get_logger();
    $logger->logdie("error: one of the parameters 'listOfStemmedTaggedSentences' or 'listOfText' must be defined.");
  }

  # set the parameter to use the original or stemmed word.
  my $useStemmedWords = 1;
  $useStemmedWords = $Parameters{useStemmedWords} if exists $Parameters{useStemmedWords};
  my $tokenIndex;
  if ($useStemmedWords) { $tokenIndex = Text::StemTagPOS::WORD_STEMMED; }
  else { $tokenIndex = Text::StemTagPOS::WORD_ORIGINAL; }

  # set the addEdgesSpanningLists option via the addEdgesSpanningSentences flag.
  $Parameters{addEdgesSpanningLists} = 1;
  $Parameters{addEdgesSpanningLists} = $Parameters{addEdgesSpanningSentences} if exists $Parameters{addEdgesSpanningSentences};

  # filter the tagged text down to only the parts-of-speech that are to be kept.
  my $listOfFilteredSentences = $Self->{posTaggerStemmerEngine}->getTaggedTextToKeep (listOfStemmedTaggedSentences => $listOfStemmedTaggedSentences);

  # build the list of sentences containing only the stemmed words kept.
  my @listOfTokens;
  foreach my $sentence (@$listOfFilteredSentences)
  {
    # skip empty sentences.
    next unless ($#$sentence + 1);

    # use only the stemmed word as the token.
    push @listOfTokens, [map {lc $_->[$tokenIndex]} @$sentence];
  }

  # get the textrank of the tokens.
  my $hashOfTextrankValues = getTextrankOfListOfTokens (%Parameters, listOfTokens => \@listOfTokens);

  # store the tagged text, filtered text, and textrank values in a hash.
  # all this info is needed to build the keywords and phrases.
  my %textrankInfo;
  $textrankInfo{listOfStemmedTaggedSentences} = $listOfStemmedTaggedSentences;
  $textrankInfo{listOfFilteredSentences} = $listOfFilteredSentences;
  $textrankInfo{hashOfTextrankValues} = $hashOfTextrankValues;
  $textrankInfo{useStemmedWords} = $useStemmedWords;

  return \%textrankInfo;
}


=head1 INSTALLATION

To install the module run the following commands:

  perl Makefile.PL
  make
  make test
  make install

If you are on a windows box you should use 'nmake' rather than 'make'.

=head1 BUGS

Please email bugs reports or feature requests to C<bug-text-categorize-textrank-en@rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Text-Categorize-Textrank-En>.  The author
will be notified and you can be automatically notified of progress on the bug fix or feature request.

=head1 AUTHOR

 Jeff Kubina<jeff.kubina@gmail.com>

=head1 COPYRIGHT

Copyright (c) 2009 Jeff Kubina. All rights reserved.
This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=head1 KEYWORDS

categorize, english, keywords, keyprhases, nlp, pagerank, textrank

=head1 SEE ALSO

=begin html

This package implements the Textrank algorithm from the report
<a href="http://bit.ly/akSJok">TextRank: Bringing Order into Texts</a>
by <a href="http://www.cse.unt.edu/~rada/">Rada Mihalcea</a> and <a href="www.cse.unt.edu/~tarau/">Paul Tarau</a>;
which is related to <a href="http://en.wikipedia.org/wiki/PageRank">pagerank</a>.

See the Lingua::EN::Tagger <a href="http://cpansearch.perl.org/src/ACOBURN/Lingua-EN-Tagger-0.15/README">README</a>
file for a list of the part-of-speech tags.

=end html

L<Lingua::EN::Tagger>, L<Lingua::Stem::Snowball>,  L<Log::Log4perl>, L<Text::Categorize::Textrank>, L<Text::StemTagPOS>

=cut

1;
# The preceding line will help the module return a true value
